docs/wiki/development/kernel.md

```diff
@@ -1,28 +1,47 @@
+
+This code is mostly shared between BSD-based kernels and Linux. The ring buffer uses spin locks to reserve structure blocks and synchronize simple writes. The minimum and maximum block reads are synchronized and reserved using an `ioctl` API. Each platform uses respective APIs to register callback methods that implement the ring buffer reserve, copy, and write.
+
+The kernel applies calling-process ownership limitations to super users. Only 1 process should issues IOCTL commands, if another pid interrupts the queue and buffer are consider invalid and all pointers reset. Clear tear down assures deregistration of callback functions and will result in maximum performance. Improper tear down may trigger timeouts and in the worst scenario continue to track callbacks.
 
-Working in the kernel directory:
+# Apple OS X Kernel Extensions
 
-First disable signature verification on the machine by running `make disable-signing` a reboot is required for this to take effect.  CAUTION: this lowers the security of your system.  It is better to have your kernel extensions signed.
+Kernel extension loading requires extension bundles signed by valid Apple developer certificates. During development it is NOT recommended to sign inline with building and running unit/integration tests. Changes to the kernel-mode code may result in kernel panics. Some of the development-only-enabled unit test code is designed to stress the limits of kernel memory management and event synchronization. It is very highly recommended to restrict these tests to running inside a virtual environment.
 
-Once kernel extension signature verification is disabled, just `make` the kernel extension.  It can then be loaded and unloaded with `make load` and `make unload`.  For permanent installations it should be added to the systems startup kernel extensions.
+The high-level kernel development workflow involves:
+- Install a copy of OS X in VMware and share a development/source/build folder.
+- Compile and build the user-land and kernel-land code on your host.
+- Load the kernel extension in the VM and run the kernel-specific unit tests.
+- You may optionally configure the VM to break on boot to debug with `lldb`.
+
+## Building on OS X for local loading (not recommended)
+
+First disable signature verification on the machine by running `make kernel-provision` a reboot is required for this to take effect.  **CAUTION:** this lowers the security of your system.  It is better to have your kernel extensions signed on your host and develop/load in a virtual machine.
+
+```
+make kernel-build
+make kernel-load
+# Optionally run the unit tests
+# make kernel-test
```

Comment:
`kernel-test` will require sudo and will assume the kernel module/extension is loaded.